Source/WebGPU/WebGPU/UsdModelRenderer.swift

```diff
@@ -0,0 +1,439 @@
+
+    @objc
+    override init() {
+        super.init()
+    }
+
+    @objc(setDeviceWithDevice:)
+    func setDeviceWith(_ device: MTLDevice) {
+        self.device = device
+
+        #if canImport(DirectDrawBackend)
+        context = DirectDrawBackend.Context.createContext(device: device)
+        scene = self.context?.makeScene()
+        commandQueue = device.makeCommandQueue()
+        captureManager = MTLCaptureManager.shared()
+        camera = context?.makeCamera()
+        context?.setCameraDistance(171.0)
+        context?.setEnableModelRotation(true)
+        #endif
+    }
+
+    @objc(renderWithTexture:)
+    func render(with texture: MTLTexture) {
+        let captureDescriptor = MTLCaptureDescriptor(from: self.device)
+        do {
+            try captureManager?.startCapture(with: captureDescriptor)
+        } catch {
+            Helper.logger.info("failed to start gpu capture")
+        }
+
+        #if canImport(DirectDrawBackend)
+
+        self.context?.update(deltaTime: 1.0 / 60.0)
+        self.camera?.renderTarget = .texture(texture)
+        guard let scene else {
+            return
+        }
+        guard let camera else {
+            return
+        }
+        self.context?.render(scene: scene, camera: camera)
+
+        captureManager?.stopCapture()
+
+        #endif
+    }
+
+    @objc(addMesh:identifier:)
+    func addMesh(_ descriptor: DDBridgeAddMeshRequest, identifier: UUID) -> Bool {
+        #if canImport(DirectDrawBackend)
+        let meshDescriptor = descriptor.reDescriptor
+        guard let context else {
+            return false
+        }
+
+        do {
+            let mesh = try context.makeMeshResource(descriptor: meshDescriptor)
+            let meshInstance = context.makeMeshInstance(
+                resource: mesh,
+                materials: [],
+                transform: matrix_identity_float4x4,
+                instanceTransforms: []
+            )
+            meshes[identifier] = meshInstance
+            guard let scene else {
+                return false
+            }
+            scene.meshes.append(meshInstance)
+            return true
+        } catch {
+            Helper.logger.info("unexpected error")
+            return false
+        }
+        #else
+        return false
+        #endif
+    }
+
+    @objc(updateMesh:identifier:)
+    func updateMesh(_ request: DDBridgeUpdateMesh, identifier: UUID) {
+        #if canImport(DirectDrawBackend)
+        guard let mesh = meshes[identifier] else {
+            Helper.logger.info("invalid mesh id")
+            return
+        }
+
+        let resource = mesh.resource
+        guard let context else {
+            return
+        }
+        do {
+            if let materialIds = request.materialIds {
+                try mesh.materials = materialIds.map { materialId in
+                    guard let m = materials[materialId] else {
+                        throw NSError()
```

Comment:
throwing errors should not be used for control flow; this should probably just be a compactMap